File: 2019/go/2.go
package _go

import (
	"fmt"
	"strconv"
	"strings"
)

type opcodefunc = func([]int, int) int

type Day2 struct{}

var ops = map[int]opcodefunc{
	1: func(buf []int, i int) int {
		buf[buf[i+3]] = buf[buf[i+1]] + buf[buf[i+2]]
		return i + 4
	},
	2: func(buf []int, i int) int {
		buf[buf[i+3]] = buf[buf[i+1]] * buf[buf[i+2]]
		return i + 4
	},
	99: func(buf []int, i int) int {
		return -1
	},
}

func (d Day2) part1(data []int, n, v int) int {
	i := 0
	buf := make([]int, len(data))
	copy(buf, data)
	buf[1] = n
	buf[2] = v
	for i != -1 {
		i = ops[buf[i]](buf, i)
	}
	return buf[0]
}

func (d Day2) part2(data []int) int {
	for n := 0; n < 100; n++ {
		for v := 0; v < 100; v++ {
			if d.part1(data, n, v) == 19690720 {
				return (100 * n) + v
			}
		}
	}
	return -1
}

func (d Day2) parse(file string) (ops []int) {
	lines := ParseFile(file)
	for _, opstr := range strings.Split(lines[0], ",") {
		op, _ := strconv.Atoi(opstr)
		ops = append(ops, op)
	}
	return
}

func (d Day2) Go() {
	buf := d.parse("in/day2")
	fmt.Println(d.part1(buf, 12, 2))
	fmt.Println(d.part2(buf))
}

File: 2019/py/2.py
#!/usr/bin/env python3
import math

filepath = "in/2"
f = open(filepath)
lines = f.readlines()
f.close()
buf = [int(x) for x in lines[0].split(",")]


def add(buf, i):
    buf[buf[i + 3]] = buf[buf[i + 1]] + buf[buf[i + 2]]
    return i + 4


def mul(buf, i):
    buf[buf[i + 3]] = buf[buf[i + 1]] * buf[buf[i + 2]]
    return i + 4


def hlt(buf, i):
    return None


ops = {1: add, 2: mul, 99: hlt}


def part1(data, n, v):
    buf = data.copy()
    buf[1] = n
    buf[2] = v
    i = 0
    while i != None:
        i = ops[buf[i]](buf, i)

    return buf[0]


def part2(data):
    for n in range(99):
        for v in range(99):
            if part1(data, n, v) == 19690720:
                return (100 * n) + v


print(part1(buf, 12, 2))  # 3306701
print(part2(buf))  # 7621
